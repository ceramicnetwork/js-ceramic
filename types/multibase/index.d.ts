declare module 'multibase' {
  export function encode(base: string, payload: Uint8Array): Uint8Array
  export function decode(val: string): Uint8Array
  export function isEncoded(bytes: Uint8Array | string): string
}
