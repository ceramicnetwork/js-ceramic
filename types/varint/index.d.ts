declare module "varint" {
  export const encode: {
    (num: number, array?: Uint8Array, offset?: number): Uint8Array;
    bytes: number;
  }
  export const decode: {
    (buf: Uint8Array, offset?: number): number;
    bytes: number;
  }
}
