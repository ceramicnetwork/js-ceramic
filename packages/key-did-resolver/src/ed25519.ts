import { toString } from 'uint8arrays/to-string'
import { convertPublicKeyToX25519 } from '@stablelib/ed25519'

function encodeKey(key: Uint8Array): string {
  const bytes = new Uint8Array(key.length + 2)
  bytes[0] = 0xec
  // The multicodec is encoded as a varint so we need to add this.
  // See js-multicodec for a general implementation
  bytes[1] = 0x01
  bytes.set(key, 2)
  return `z${toString(bytes, 'base58btc')}`
}

/**
 * Constructs the document based on the method key
 */
export function keyToDidDoc(pubKeyBytes: Uint8Array, fingerprint: string): any {
  const did = `did:key:${fingerprint}`
  const keyId = `${did}#${fingerprint}`
  const x25519PubBytes = convertPublicKeyToX25519(pubKeyBytes)
  const x25519KeyId = `${did}#${encodeKey(x25519PubBytes)}`
  return {
    id: did,
    verificationMethod: [
      {
        id: keyId,
        type: 'Ed25519VerificationKey2018',
        controller: did,
        publicKeyBase58: toString(pubKeyBytes, 'base58btc'),
      },
    ],
    authentication: [keyId],
    assertionMethod: [keyId],
    capabilityDelegation: [keyId],
    capabilityInvocation: [keyId],
    keyAgreement: [
      {
        id: x25519KeyId,
        type: 'X25519KeyAgreementKey2019',
        controller: did,
        publicKeyBase58: toString(x25519PubBytes, 'base58btc'),
      },
    ],
  }
}
