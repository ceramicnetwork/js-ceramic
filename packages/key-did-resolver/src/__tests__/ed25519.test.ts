import varint from 'varint'
import multibase from 'multibase'
import * as mapper from '../ed25519'

describe('Ed25519 mapper', () => {
  it('successfully resolves the document from did', async () => {
    const id = 'z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8'

    const multicodecPubKey = multibase.decode(id)
    varint.decode(multicodecPubKey) // decode is changing param multicodecPubKey as well
    const pubKeyBytes = multicodecPubKey.slice(varint.decode.bytes)
    const doc = await mapper.keyToDidDoc(pubKeyBytes, id)
    expect(doc).toMatchSnapshot()
  })
})
