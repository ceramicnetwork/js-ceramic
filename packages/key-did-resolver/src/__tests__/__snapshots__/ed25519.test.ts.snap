// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ed25519 mapper successfully resolves the document from did 1`] = `
Object {
  "assertionMethod": Array [
    "did:key:z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8#z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8",
  ],
  "authentication": Array [
    "did:key:z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8#z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8",
  ],
  "capabilityDelegation": Array [
    "did:key:z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8#z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8",
  ],
  "capabilityInvocation": Array [
    "did:key:z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8#z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8",
  ],
  "id": "did:key:z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8",
  "keyAgreement": Array [
    Object {
      "controller": "did:key:z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8",
      "id": "did:key:z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8#z6LSnkZe3JZPCo88XsdQVJi8j1TomzX2yRW7ZnvvWhmrSdmG",
      "publicKeyBase58": "C5PUWzkX7LQPSVFdxfCBQRFKvqyvGpKxgpDF2F8KjFzW",
      "type": "X25519KeyAgreementKey2019",
    },
  ],
  "verificationMethod": Array [
    Object {
      "controller": "did:key:z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8",
      "id": "did:key:z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8#z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8",
      "publicKeyBase58": "FUaAP6i2XyyouPds73QneYgZJ86qhua2jaZYBqJSwKok",
      "type": "Ed25519VerificationKey2018",
    },
  ],
}
`;
